app/code/Magento/Catalog/Model/ProductRepository.php

```diff
@@ -302,12 +302,14 @@ protected function getCacheKey($data)
      *
      * @param string $cacheKey
      * @param \Magento\Catalog\Api\Data\ProductInterface $product
+     * @param null|string $sku
      * @return void
      */
-    private function cacheProduct($cacheKey, \Magento\Catalog\Api\Data\ProductInterface $product)
+    private function cacheProduct($cacheKey, \Magento\Catalog\Api\Data\ProductInterface $product, $sku = null)
```

Comment:
See: https://github.com/magento/magento2/commit/cb3b5d1e0715632c15d8a9697538afca4083a859#diff-ff4893a5628d34910295bae32fbcddfdL241

Please pass `productId` instead of product instance in both cases. Don't add default parameter and some logic into `cacheProduct`, pass both product ID and SKU.